hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/AggregateAbstractPhysicalRule.java

```diff
@@ -70,71 +76,489 @@ protected AggregateAbstractPhysicalRule(Config config) {
                     int minIndex = aggregateCallArguments.get(0);
                     aggregationProviders.add(MinSqlAggregation::new);
-                    valueProviders.add(row -> row.get(minIndex));
+                    valueProviders.add(new GetFunction(minIndex));
                     break;
                 case MAX:
                     int maxIndex = aggregateCallArguments.get(0);
                     aggregationProviders.add(MaxSqlAggregation::new);
-                    valueProviders.add(row -> row.get(maxIndex));
+                    valueProviders.add(new GetFunction(maxIndex));
                     break;
                 case SUM:
                     int sumIndex = aggregateCallArguments.get(0);
                     QueryDataType sumOperandType = operandTypes.get(sumIndex);
-                    aggregationProviders.add(() -> SumSqlAggregations.from(sumOperandType, distinct));
-                    valueProviders.add(row -> row.get(sumIndex));
+                    aggregationProviders.add(new AggregateSumSupplier(distinct, sumOperandType));
+                    valueProviders.add(new GetFunction(sumIndex));
                     break;
                 case AVG:
                     int avgIndex = aggregateCallArguments.get(0);
                     QueryDataType avgOperandType = operandTypes.get(avgIndex);
-                    aggregationProviders.add(() -> AvgSqlAggregations.from(avgOperandType, distinct));
-                    valueProviders.add(row -> row.get(avgIndex));
+                    aggregationProviders.add(new AggregateAvgSupplier(distinct, avgOperandType));
+                    valueProviders.add(new GetFunction(avgIndex));
                     break;
                 default:
                     throw QueryException.error("Unsupported aggregation function: " + kind);
             }
         }
 
         return AggregateOperation
-                .withCreate(() -> {
-                    List<SqlAggregation> aggregations = new ArrayList<>(aggregationProviders.size());
-                    for (SupplierEx<SqlAggregation> aggregationProvider : aggregationProviders) {
-                        aggregations.add(aggregationProvider.get());
-                    }
-                    return aggregations;
-                })
-                .andAccumulate((List<SqlAggregation> aggregations, JetSqlRow row) -> {
-                    for (int i = 0; i < aggregations.size(); i++) {
-                        aggregations.get(i).accumulate(valueProviders.get(i).apply(row));
-                    }
-                })
-                .andCombine((lefts, rights) -> {
-                    assert lefts.size() == rights.size();
+                .withCreate(new AggregateCreateSupplier(aggregationProviders))
+                .andAccumulate(new AggregateAccumulateFunction(valueProviders))
+                .andCombine(AggregateCombineFunction.INSTANCE)
+                .andExportFinish(AggregateExportFinishFunction.INSTANCE);
```

Comment:
We should order these four classes in the file according to the order appearing here.

Reply:
Done